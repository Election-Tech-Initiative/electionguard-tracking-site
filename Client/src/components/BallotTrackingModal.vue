<template>
  <div class="modal" :class="{ 'is-active': isActive }">
    <div class="modal-background" @click="$emit('close')"></div>
    <div class="modal-content">
      <div class="box">
        <LoadingIndicator v-if="isLoading" />
        <div v-else-if="selectedBallot"
          :class="{
            'counted': selectedBallot.status === 0,
            'not-found': selectedBallot.status === 404,
            'discarded': selectedBallot.status === -1, // assuming status of -1 for now
          }">

          <h1 class="title">
            <span v-if="selectedBallot.status === 0">{{ $t('ballotCountedTitle') }}</span>
            <span v-else-if="selectedBallot.status === 404">{{ $t('ballotNotFoundTitle') }}</span>
            <span v-else-if="selectedBallot.status === -1">{{ $t('ballotDiscardedTitle') }}</span>
          </h1>

          <div class="content">
            <div v-if="selectedBallot.status === 0">
              <p>{{ $t('ballotCountedTopMessage', {location: appSettings.locationDescription}) }}</p>
            </div>
            <div v-else-if="selectedBallot.status === 404">
              <p>{{ $t('ballotNotFoundTopMessage', {location: appSettings.locationDescription}) }}</p>
              <ol>
                <li>{{ $t('ballotNotFoundListItem1') }}</li>
                <li>
                  <i18n path="ballotNotFoundListItem2">
                    <a :href="appSettings.commissionContact">{{ $t('ballotNotFoundListItem2LinkLabel') }}</a>
                  </i18n>
                </li>
              </ol>
            </div>
            <div v-else-if="selectedBallot.status === -1">
              <p>{{ $t('ballotDiscardedTopMessage', {location: appSettings.locationDescription}) }}</p>
            </div>
          </div>

          <div class="content ballot-status">
            <div class="ballot-data has-text-weight-semibold">
              {{selectedBallot.trackingId}}
            </div>
            <div v-if="selectedBallot.status === 0 || selectedBallot.status === -1"
              class="ballot-status-bottom">
              <div class="ballot-data">
                <span class="has-text-weight-semibold">{{ $t('ballotLocationLabel') }}</span> {{selectedBallot.location}}
              </div>
              <div class=ballot-data>
                <span class="has-text-weight-semibold">{{ $t('ballotVoteTimeLabel') }}</span> {{approximateCastTime}}
              </div>
            </div>
          </div>
          
          <div class="content">
            <div v-if="selectedBallot.status === 0">
              <i18n tag="p" path="ballotCountedBottomMessage">
                <a href="https://aka.ms/ElectionGuardMoreInfo">{{ $t('ballotClickHere') }}</a>
              </i18n>
            </div>
            <div v-else-if="selectedBallot.status === 404">
              <i18n tag="p" path="ballotNotFoundBottomMessage">
                <a href="https://aka.ms/ElectionGuardMoreInfo">{{ $t('ballotClickHere') }}</a>
              </i18n>
            </div>
            <div v-else-if="selectedBallot.status === -1">
              <i18n tag="p" path="ballotDiscardedBottomMessage">
                <a href="https://aka.ms/ElectionGuardMoreInfo">{{ $t('ballotClickHere') }}</a>
              </i18n>
            </div>
          </div>
          
          <LanguageToggle />
          <a class="button is-dark has-text-weight-bold" href="#" @click.prevent="$emit('close')">
            {{ $t('close') }}
          </a>

        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';
import moment from 'moment-timezone';

import { BallotStatus } from '@/models/BallotStatus';
import LanguageToggle from '@/components/LanguageToggle.vue';
import LoadingIndicator from '@/components/LoadingIndicator.vue';

@Component({
  components: {
    LanguageToggle,
    LoadingIndicator,
  },
  props: {
    isActive: {
      type: Boolean
    }
  }
})
export default class BallotTrackingModal extends Vue {
  @Prop() public isActive?: boolean;
  @Prop() public isLoading?: boolean;
  @Prop() public selectedBallot?: BallotStatus;

  private readonly MinuteInterval = 5;

  public get approximateCastTime(): string {
    if (!!this.selectedBallot) {
      const castTime = moment.utc(this.selectedBallot.approximateCastTime);
      const minutes = castTime.minutes();
      const roundedMins = Math.round(minutes / this.MinuteInterval) * this.MinuteInterval;
      const approximateDate = castTime.minutes(roundedMins);
      const tz = moment.tz.guess();
      const date = approximateDate.local().tz(tz).format('MMMM D, YYYY');
      return date;
    }
    return '';
  }
}
</script>

<style lang="scss" scoped>
.language-toggle {  
    float: right;
    margin-bottom: 1.5rem;
}

.box {
  border-radius: 0.2rem;
  padding: 2rem;

  @include mobile {
    padding: 1.5rem;
  }
}

.button.is-dark {
  font-size: 0.9rem;
  padding: 1.2rem 2rem;
}

@include tablet {
  .modal-content {
    width: 75%;
    max-width: 1080px;
  }
}

// default colors to NOT FOUND version of design
h1.title {
  display: flex;
  align-items: center;
  color: $primary;

  &::before {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    content: '?';
    color: $white;
    font-size: 2.5rem;
    background-color: $primary;
    border-radius: 50%;
    width: 4rem;
    height: 4rem;
    margin-right: 1.5rem;
  }

  @include mobile {
    font-size: 1.4rem;

    &::before {
      font-size: 2rem;
      margin-right: 1rem;
      width: 3rem;
      height: 3rem;
    }
  }
}

.ballot-status {
  > div {
    font-size: 0.9rem;
    border: 0.1rem solid $primary;
    padding: 1rem 1.1rem 0 1.1rem;
  }

  .ballot-status-bottom {
    border-top: none;
  }

  .ballot-data {
    padding-bottom: 1rem;
  }
}

.counted {
  h1.title {
    color: $success;

    &::before {
      content: '\2713';
      background-color: $success;
    }
  }

  .ballot-status {
    > div {
      border-color: $success;
    }
  }
}

.discarded {
  h1.title {
    color: $danger;
    &::before {
      content: 'X';
      background-color: $danger;
    }
  }

  .ballot-status {
    > div {
      border-color: $danger;
    }
  }
}
</style>
