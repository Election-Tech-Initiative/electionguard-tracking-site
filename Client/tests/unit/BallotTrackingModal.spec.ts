import { mount } from '@vue/test-utils';

import BallotTrackingModal from '@/components/BallotTrackingModal.vue';

jest.mock('@/components/LanguageToggle.vue', () => ({ default: jest.fn() }));
jest.mock('@/components/LoadingIndicator.vue', () => ({ default: jest.fn() }));

describe('BallotTrackingModal.vue', () => {
    const getMountedWrapper = (status: number) => mount(BallotTrackingModal, {
        mocks: {
            appSettings: {
                commissionContact: 'Mocked Contact',
                locationDescription: 'Mocked Location',
            },
            $t: (msg: string): string => msg
        },
        propsData: {
            isActive: true,
            isLoading: false,
            selectedBallot: {
                approximateCastTime: new Date('July 31, 2019 12:30:00'),
                location: 'mockedLocation',
                status,
                trackingId: 'mockedTrackingId',
            },
        },
        stubs: [
            'i18n',
            'LanguageToggle',
            'LoadingIndicator',
        ],
    });

    it('renders for ballot counted', () => {
        const wrapper = getMountedWrapper(0);
        expect(wrapper.element).toMatchSnapshot('for ballot counted');
    });

    it('renders for ballot not found', () => {
        const wrapper = getMountedWrapper(404);
        expect(wrapper.element).toMatchSnapshot('for ballot not found');
    });

    it('renders for ballot discarded', () => {
        const wrapper = getMountedWrapper(-1);
        expect(wrapper.element).toMatchSnapshot('for ballot discarded');
    });
});
