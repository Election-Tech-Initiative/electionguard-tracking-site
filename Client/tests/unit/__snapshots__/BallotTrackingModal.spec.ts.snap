// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BallotTrackingModal.vue renders for ballot counted: for ballot counted 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
  />
   
  <div
    class="modal-content"
  >
    <div
      class="box"
    >
      <div
        class="counted"
      >
        <h1
          class="title"
        >
          <span>
            ballotCountedTitle
          </span>
        </h1>
         
        <div
          class="content"
        >
          <div>
            <p>
              ballotCountedTopMessage
            </p>
          </div>
        </div>
         
        <div
          class="content ballot-status"
        >
          <div
            class="ballot-data has-text-weight-semibold"
          >
            
            mockedTrackingId
          
          </div>
           
          <div
            class="ballot-status-bottom"
          >
            <div
              class="ballot-data"
            >
              <span
                class="has-text-weight-semibold"
              >
                ballotLocationLabel
              </span>
               mockedLocation
            
            </div>
             
            <div
              class="ballot-data"
            >
              <span
                class="has-text-weight-semibold"
              >
                ballotVoteTimeLabel
              </span>
               July 31, 2019 12:30 pm EDT
            
            </div>
          </div>
        </div>
         
        <div
          class="content"
        >
          <div>
            <i18n-stub
              path="ballotCountedBottomMessage"
              tag="p"
            >
              <a
                href="https://aka.ms/ElectionGuardMoreInfo"
              >
                ballotClickHere
              </a>
            </i18n-stub>
          </div>
        </div>
         
        <languagetoggle-stub />
         
        <a
          class="button is-dark has-text-weight-bold"
          href="#"
        >
          
          close
        
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`BallotTrackingModal.vue renders for ballot discarded: for ballot discarded 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
  />
   
  <div
    class="modal-content"
  >
    <div
      class="box"
    >
      <div
        class="discarded"
      >
        <h1
          class="title"
        >
          <span>
            ballotDiscardedTitle
          </span>
        </h1>
         
        <div
          class="content"
        >
          <div>
            <p>
              ballotDiscardedTopMessage
            </p>
          </div>
        </div>
         
        <div
          class="content ballot-status"
        >
          <div
            class="ballot-data has-text-weight-semibold"
          >
            
            mockedTrackingId
          
          </div>
           
          <div
            class="ballot-status-bottom"
          >
            <div
              class="ballot-data"
            >
              <span
                class="has-text-weight-semibold"
              >
                ballotLocationLabel
              </span>
               mockedLocation
            
            </div>
             
            <div
              class="ballot-data"
            >
              <span
                class="has-text-weight-semibold"
              >
                ballotVoteTimeLabel
              </span>
               July 31, 2019 12:30 pm EDT
            
            </div>
          </div>
        </div>
         
        <div
          class="content"
        >
          <div>
            <i18n-stub
              path="ballotDiscardedBottomMessage"
              tag="p"
            >
              <a
                href="https://aka.ms/ElectionGuardMoreInfo"
              >
                ballotClickHere
              </a>
            </i18n-stub>
          </div>
        </div>
         
        <languagetoggle-stub />
         
        <a
          class="button is-dark has-text-weight-bold"
          href="#"
        >
          
          close
        
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`BallotTrackingModal.vue renders for ballot not found: for ballot not found 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
  />
   
  <div
    class="modal-content"
  >
    <div
      class="box"
    >
      <div
        class="not-found"
      >
        <h1
          class="title"
        >
          <span>
            ballotNotFoundTitle
          </span>
        </h1>
         
        <div
          class="content"
        >
          <div>
            <p>
              ballotNotFoundTopMessage
            </p>
             
            <ol>
              <li>
                ballotNotFoundListItem1
              </li>
               
              <li>
                <i18n-stub
                  path="ballotNotFoundListItem2"
                >
                  <a
                    href="Mocked Contact"
                  >
                    ballotNotFoundListItem2LinkLabel
                  </a>
                </i18n-stub>
              </li>
            </ol>
          </div>
        </div>
         
        <div
          class="content ballot-status"
        >
          <div
            class="ballot-data has-text-weight-semibold"
          >
            
            mockedTrackingId
          
          </div>
           
          <!---->
        </div>
         
        <div
          class="content"
        >
          <div>
            <i18n-stub
              path="ballotNotFoundBottomMessage"
              tag="p"
            >
              <a
                href="https://aka.ms/ElectionGuardMoreInfo"
              >
                ballotClickHere
              </a>
            </i18n-stub>
          </div>
        </div>
         
        <languagetoggle-stub />
         
        <a
          class="button is-dark has-text-weight-bold"
          href="#"
        >
          
          close
        
        </a>
      </div>
    </div>
  </div>
</div>
`;
